pipelines/batch/src/main/java/org/openmrs/analytics/JdbcFetchHapi.java

```diff
@@ -0,0 +1,146 @@
+		
+		public FetchRowsJdbcIo(JdbcIO.DataSourceConfiguration dataSourceConfig) {
+			this.dataSourceConfig = dataSourceConfig;
+		}
+		
+		@Override
+		public PCollection<JdbcHapiRowDescriptor> expand(PCollection<List<String>> queryParameters) {
+			return queryParameters.apply("JdbcIO readAll",
+			    JdbcIO.<List<String>, JdbcHapiRowDescriptor> readAll().withDataSourceConfiguration(dataSourceConfig)
+			            .withParameterSetter(new JdbcIO.PreparedStatementSetter<List<String>>() {
+				            
+				            @Override
+				            public void setParameters(List<String> element, PreparedStatement preparedStatement)
+				                    throws Exception {
+					            preparedStatement.setString(1, element.get(0));
+					            preparedStatement.setInt(2, Integer.valueOf(element.get(1)));
+					            preparedStatement.setInt(3, Integer.valueOf(element.get(2)));
+				            }
+			            })
+			            // We are disabling this parameter because by default, this parameter causes JdbcIO to add a 
+			            // reshuffle transform after reading from the database. This breaks fushion between the read 
+			            // and write operations, thus resulting in high memory overhead. Diabling the below parameter 
+			            // results in optimal performance.
+			            .withOutputParallelization(false)
+			            .withQuery(
+			                "SELECT res.res_id, res.res_type, res.res_updated, res.res_ver, ver.res_encoding, ver.res_text "
+			                        + "FROM hfj_resource res, hfj_res_ver ver "
+			                        + "WHERE res.res_type=? AND res.res_id = ver.res_id AND res.res_id %? = ?")
+			            .withRowMapper(new ResultSetToRowDescriptor()));
+			
+		}
+	}
+	
+	JdbcIO.DataSourceConfiguration getJdbcConfig() {
+		return JdbcIO.DataSourceConfiguration.create(this.jdbcConnectionUtil.getConnectionObject());
+	}
+	
+	/**
+	 * Generates the query parameters for the JdbcIO fetch. The query parameters are generated based on
+	 * the batch size option set for the job.
+	 * 
+	 * @param options the pipeline options
+	 * @param resourceType the resource type
+	 * @param numResources total number of resources of the given type
+	 * @param batchNum the relevant batch number
+	 * @return a list of query parameters in the form of lists of strings
+	 */
+	@VisibleForTesting
+	List<List<String>> generateQueryParameters(FhirEtlOptions options, String resourceType, int numResources) {
+		log.info("Generating query parameters for " + resourceType);
+		
+		int jdbcFetchSize = options.getJdbcFetchSize();
+		List<List<String>> queryParameterList = new ArrayList<List<String>>();
+		int numBatches = numResources / jdbcFetchSize;
+		if (numResources % jdbcFetchSize != 0) {
+			numBatches += 1;
+		}
+		
+		for (int i = 0; i < numBatches; i++) {
+			queryParameterList.add(Arrays.asList(resourceType, String.valueOf(numBatches), String.valueOf(i)));
```

Comment:
Beside readability, I think you can avoid casting everything to `String` if you move away from `List` to a normal data object.

Reply:
Changes made.